rust/lance/src/io/commit.rs

```diff
@@ -253,21 +254,13 @@ async fn migrate_indices(dataset: &Dataset, indices: &mut [Index]) -> Result<()>
 
     for index in indices {
         if index.fragment_bitmap.is_none() {
-            // Load the read version of the index
-            let old_version = match dataset.checkout_version(index.dataset_version).await {
-                Ok(dataset) => dataset,
-                // If the version doesn't exist anymore, skip it.
-                Err(crate::Error::DatasetNotFound { .. }) => continue,
-                // Any other error we return.
-                Err(e) => return Err(e),
-            };
-            index.fragment_bitmap = Some(
-                old_version
-                    .get_fragments()
-                    .iter()
-                    .map(|f| f.id() as u32)
-                    .collect(),
-            );
+            debug_assert_eq!(index.fields.len(), 1);
+            let idx_field = dataset.schema().field_by_id(index.fields[0]).ok_or_else(|| Error::Internal { message: format!("Index with uuid {} referred to field with id {} which did not exist in dataset", index.uuid, index.fields[0]), location: location!() })?;
+            // We need to calculate the fragments covered by the index
+            let idx = dataset
+                .open_generic_index(&idx_field.name, &index.uuid.to_string())
+                .await?;
+            index.fragment_bitmap = Some(idx.calculate_included_frags().await?);
```

Comment:
Perhaps we should have a fast path, where we see if [index.dataset_version](https://github.com/lancedb/lance/blob/48b19829d3c7bd07ebadb3ebc1c3b7eb2b6c9068/protos/index.proto#L37C16-L37C16) still exists and if it does load the fragments from that manifest? Does that seem like it would be faster than loading from the index?

Reply:
Loading the fragments from that manifest could be incorrect because we update the dataset version when we optimize indices.  So we could have:

TRAIN ON DS VERSION 1 FRAGS 0,1,2,3
DELETE FRAG 0
OPTIMIZE TO DS VERSION 2 FRAGS 1, 2, 3

Now, `index.dataset_version` is `2` which points to frags `1, 2, 3` but the index actually covers frags `0, 1, 2, 3`.

Reply:
That makes sense. Out of curiosity: Do we update the UUID of the index when we optimize the index? That would let us detect whether the original fragments can be used.

I'm not sure we could use that if we could; it's just extra IOs for the tables that need fixing.

Reply:
The UUID changes anytime the index changes (compact or update/optimize) though I'm not sure how that helps?

Reply:
It provides an algorithm to find if an index has never been optimized:

1. Checkout `dataset_version`
2. Compare the `UUID` of index versus the `UUID` in `dataset_version`. If they are the same, the index has never been optimized. Otherwise, it has been optimized / compacted.

Reply:
Ok, that's pretty clever.  Though I'll probably hold off as I don't expect this recalculate to run that often and it isn't *that* expensive.

Reply:
Yeah, I agree it's not worth implementing for this issue. But something to keep in mind we *can* do, if it ever comes up again.